src/main/java/org/jitsi/jibri/api/xmpp/XmppApi.kt

```diff
@@ -0,0 +1,214 @@
+                    mucClient.sendStanza(jibriStatus)
+                }
+                mucClient.createOrJoinMuc(
+                    JidCreate.entityBareFrom("${config.controlMuc.roomName}@${config.controlMuc.domain}"),
+                    Resourcepart.from(config.controlMuc.nickname))
+
+                val jibriStatus = JibriPresenceHelper.createPresence(
+                    JibriStatusPacketExt.Status.IDLE,
+                    JidCreate.bareFrom("${config.controlMuc.roomName}@${config.controlMuc.domain}"))
+                mucClient.sendStanza(jibriStatus)
+            }
+        }
+    }
+
+    /**
+     * Helper function to handle a [JibriIq] message with the context of the [XmppEnvironmentConfig] and [MucClient]
+     * that this [JibrIq] was received on.
+     */
+    private fun handleJibriIq(jibriIq: JibriIq, xmppEnvironment: XmppEnvironmentConfig, mucClient: MucClient): IQ {
+        logger.info("Received JibriIq $jibriIq from environment ${xmppEnvironment.name}")
+        return when (jibriIq.action) {
+            JibriIq.Action.START -> handleStartJibriIq(jibriIq, xmppEnvironment, mucClient)
+            JibriIq.Action.STOP -> handleStopJibriIq(jibriIq)
+            else -> IQ.createErrorResponse(
+                jibriIq,
+                XMPPError.getBuilder().setCondition(XMPPError.Condition.bad_request))
+        }
+    }
+
+    /**
+     * Handle a start [JibriIq] message.  We'll respond immediately with a [JibriIq.Status.PENDING] IQ response and
+     * send a new IQ with the subsequent stats after starting the service:
+     * [JibriIq.Status.FAILED] if there was an error starting the service (or an error while the service was running)
+     * [JibriIq.Status.BUSY] if the Jibri was already busy and couldn't start the requested service
+     * [JibriIq.Status.ON] if the service started successfully
+     */
+    private fun handleStartJibriIq(
+            startJibriIq: JibriIq,
+            xmppEnvironment: XmppEnvironmentConfig,
+            mucClient: MucClient): IQ {
+        logger.info("Received start request")
+        // We don't want to block the response to wait for the service to actually start, so submit a job to
+        // start the service asynchronously and send an IQ with the status after its done.
+        executor.submit {
+            val resultIq = JibriIq()
+            resultIq.to = startJibriIq.from
+            resultIq.type = IQ.Type.set
+            try {
+                logger.info("Starting service")
+
+                // If there is an issue with the service while it's running, we need to send an IQ
+                // to notify the caller who invoked the service of its status, so we'll listen
+                // for the service's status while it's running and this method will be invoked
+                // if it changes
+                val serviceStatusHandler: JibriServiceStatusHandler = { serviceStatus ->
```

Comment:
I think this was the function with the 9 cyclomatic complexity. Just based on the level of nesting, and the normal level of nesting you've been aiming for, I think this one stands out as a "beast" compared to the lightweight helpers you've been creating. I could see the function being broken out into a separate helper (and thereby potentially becoming easier to test) and maybe...enum/mapping for the stats?

Reply:
yeah i'll probably play with this a bit.